File: solutions/java/yacht/1/src/main/java/Yacht.java
import java.util.Arrays;
import java.util.Collections;
import java.util.stream.IntStream;
class Yacht {
    private final int score;
    Yacht(int[] dice, YachtCategory yachtCategory) {
        int diceTotal = Arrays.stream(dice).sum();
        switch (yachtCategory) {
            case ONES:
            case TWOS:
            case THREES:
            case FOURS:
            case FIVES:
            case SIXES:
                int categoryNum = yachtCategory.ordinal();
                score = (int) Arrays.stream(dice).filter(i -> i == categoryNum).count() * categoryNum;
                break;
            case FULL_HOUSE:
                score = matchesFrequency(dice, 3, true) && matchesFrequency(dice, 2, true) ? diceTotal : 0;
                break;
            case FOUR_OF_A_KIND:
                score = matchesFrequency(dice, 4, false) ? 4 * numRepeatedFourTimes(dice) : 0;
                break;
            case LITTLE_STRAIGHT:
                score = straightResult(dice, 1);
                break;
            case BIG_STRAIGHT:
                score = straightResult(dice, 2);
                break;
            case CHOICE:
                score = diceTotal;
                break;
            default:
                score = matchesFrequency(dice, 5, true) ? 50 : 0;
        }
    }

    int score() {
        return score;
    }

    private boolean matchesFrequency(int[] dice, int count, boolean countExact) {
        return Arrays.stream(dice).anyMatch(i -> {
            int currentCount = (int) Arrays.stream(dice).filter(j -> j == i).count();
            return countExact ? currentCount == count : currentCount >= count;
        });
    }

    private int numRepeatedFourTimes(int[] dice) {
        return dice[0] == dice[1] ? dice[0] : dice[2];
    }

    private int straightResult(int[] dice, int start) {
        return IntStream.range(start, start + 5).allMatch(i -> Arrays.stream(dice).anyMatch(j -> j == i)) ? 30 : 0;
    }
}

File: solutions/python/yacht/1/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11

DICE_COUNT = 5


def score(dice: list[int], category: int) -> int:
    if category == YACHT:
        return 50 if same_face(dice) else 0
    if ONES <= category <= SIXES:
        return category * dice.count(category)
    
    dice.sort()
    
    if category == FULL_HOUSE:
        return sum(dice) if (len(set(dice)) == 2
                             and (same_face(dice[:3]) and same_face(dice[3:])
                             or same_face(dice[:2]) and same_face(dice[2:]))) else 0
    if category == FOUR_OF_A_KIND:
        return dice[2]*4 if dice.count(dice[2]) >= 4 else 0
    if category == LITTLE_STRAIGHT:
        return 30 if dice == list(range(1, DICE_COUNT+1)) else 0
    if category == BIG_STRAIGHT:
        return 30 if dice == list(range(2, DICE_COUNT+2)) else 0
    if category == CHOICE:
        return sum(dice)
    return 0

def same_face(dice: list[int]) -> bool:
    return dice.count(dice[0]) == len(dice)
